### PR TITLE
WIP lerna bootstrap

### PR DESCRIPTION
**Pull request by ctaepper**
test repo: https://github.com/ctaepper/renovate-lerna

relates to #1441 

`master` branch: lerna & npm without module hoisting
`hoist` branch: lerna & npm with module hoisting
`yarn` branch: lerna & yarn / workspaces (just for the sake of completeness)

run `npm start` in project root to see the output of the project

`packages/p1` is an internal dependency of `packages/p2`. both are scope by `@myscope`

so executing `npm install` directly in `p2` yields

```
➜  p2 git:(master) npm i
npm ERR! code E404
npm ERR! 404 Not Found: @myscope/p1@*
```

### CODE REVIEW

**Comment by rarkins** [comment]
@ctaepper it's ready for your testing now on something "real"

I used your example/test repo and confirmed first that it would fail as expected using the old `npm install` approach. After adding `lerna bootstrap` awareness, that error went away, although the dependencies as you wrote them didn't need any updates to `package-lock.json`, so I added an out of date `chalk` and rolled back `lerna` to test. You can see some valid looking PRs here now: https://github.com/renovate-tests/renovate-lerrna/pulls

**Comment by ctaepper** [comment]
with real you mean, i should now be able to use this branch on one of our real projects right? will test and report results

**Comment by rarkins** [comment]
yes, try a full project, although I recommend maybe forking it first if you want to be safe. Add `renovateFork=true` to your config to get Renovate to run on a fork.

**Comment by ctaepper** [comment]
```
➜  renovate git:(feat/1441-lerna-bootstrap) node lib/renovate.js --token xxx --expose-env my/repo
 WARN: No repositories found - did you want to run with flag --autodiscover?
 INFO: Renovate finished
```

if i remove the `--expose-env` option, it does work, but cannot install due to the env issue from the other PR. but this is supposed to work already in this branch, right?

**Comment by ctaepper** [comment]
after applying the fix from #1445 renovate runs without errors
however, only the pin-dockerfile PR is created, but not the pin-dependecies PR?

**Comment by rarkins** [comment]
#1445 is now merged and updated into this branch too.

Renovate preset "config:base" defaults to maximum 2 PRs created per hour, perhaps you can check in case that is what was preventing the Pin Dependencies PR? Otherwise you need to look through the logs to guess what's going on. Even default log level might be enough detail. 

Also, any "upgrade" PRs are held back until after "pin" PRs are merged. So either disable Docker renovation if you want that PR to go away, or merge it to make sure it doesn't hold anything up.

**Comment by ctaepper** [comment]
before I start digging, does renovate detect any previously closed PR's? I tried to renovate those repos before with the github app version, before discovering that it doesn't work out of the box. so there are closed unmerged "pin dependencies" PRs in those repos

**Comment by rarkins** [comment]
Pin Pra should always come back like zombies - you can’t kill them unless you fix/update config. Upgrades will be blocked if an exact match.

**Comment by rarkins** [comment]
You can also set `recreateClosed=true` in config during testing if you want to be able to just close/delete a PR and have Renovate create it every time regardless. That way both Pin and Update PRs will all get recreated whenever necessary.

**Comment by ctaepper** [comment]
quick update: the pin-dependencies branch was not created because we previously already pinned all dependencies manually 🙄

I pulled in the current master and renovate tries to do its job, sadly the childprocess fails with the good old 404. it feels like renovate still tries to `npm i` instead of `lerna bootstrap`

will look into the _how_ any _why_ tonight

**Comment by rarkins** [comment]
Thanks for updating me. BTW a little tip from when I was testing - I made sure not to name the repository with "lerna" in it so that it's much faster to grep/search through debug logs for any occurrences of lerna in the log *messages* :)

The logic is meant to be like this:
- If `lerna.json` + `package-lock.json` in root of repository then we set a field `lernaPackageLock` to true.
- Later, when generating lockFiles we only call `npm install` if that field is not true
- Instead, we run `npm install && npx lerna bootstrap`

I actually have some doubts about the `npm install` definitely being necessary.. but it's there for now as you suggested.

File: lib/workers/branch/lerna.js [hunk 1]
```diff
@@ -65,7 +65,7 @@ async function generateLockFiles(tmpDir, env) {
       logger.warn('Could not detect lerna in devDependencies');
     }
     logger.debug('Using lerna version ' + lernaVersion);
-    const cmd = `find . && ls -l && ${npmCmd} install && ${npxCmd} lerna@${lernaVersion} bootstrap -- --package-lock-only`;
+    const cmd = `find . && ls -l && ${npmCmd} install --package-lock-only --ignore-scripts && ${npxCmd} lerna@${lernaVersion} bootstrap -- --package-lock-only --ignore-scripts`;
```
Comment (ctaepper) [inline lib/workers/branch/lerna.js]: package-lock-only was missing in at the install cmd, so node_modules where installed for root

the npx lerna command causes some issues
- it seems like it is ignoring the package-lock-only parameter (it does work if i execute the same command manually)
- it seems like the progress output of lerna is piped to stderr, which results in a "false negative" for npm install `"stderr": "npx: Installierte 295 in 7.46s\nlerna info version 2.5.1\nlerna info Bootstrapping 8 packages\nlerna info lifecycle preinstall\nlerna info Installing external dependencies\nlerna info Symlinking packages and binaries\nlerna info lifecycle postinstall\nlerna info lifecycle prepublish\nlerna info lifecycle prepare\nlerna success Bootstrapped 8 packages\n"`

I am unsure how to pervent that, any suggestions?

Comment (rarkins) [reply lib/workers/branch/lerna.js]: Is the extra `--` before ``package-lock-only` perhaps causing the problem?

I noticed the `stderr` as well, I'm not sure we can really fix that but I suppose can try to ignore it. Is it causing it to then error within the Renovate app too?

Comment (rarkins) [reply lib/workers/branch/lerna.js]: BTW when I was seeing the `stderr` messages I was still getting the lock files I was after in the PRs I  linked to

**Comment by rarkins** [comment]
@ctaepper was it just the ignore scripts causing your failures?

**Comment by ctaepper** [comment]
@rarkins it looks good now, everything seems to be working

> sadly the childprocess fails with the good old 404.

while i pulled in the current master, i overlooked to pull in your previous updates in this branch 😳 guess it was late :)

> I actually have some doubts about the npm install definitely being necessary.. but it's there for now as you suggested.

it has to be executed in order to update the root package-lock, `lerna bootstrap` only handles the mono child packages, not the root package

**Comment by ctaepper** [comment]
in my tests, postinstall wasn't executed, so i assume, that `--package-lock-only` disables scripts by default. so the `--ignore-scripts` was obsolete. `--package-lock-only` had to be added though

**Comment by rarkins** [comment]
@ctaepper that's great news. Before merging, I'll need to add tests for this new code. i can do this but probably later in the week. If you're in a hurry can you run off this branch in the meantime?

**Comment by ctaepper** [comment]
yes, no worries, I'll renovate manually until you release the new version

**Comment by rarkins** [comment]
@LucaT1 I think you may need this PR too, although updated for non-Workspaces Yarn

**Comment by ctaepper** [comment]
@rarkins thumbs up for the progress!

I ran into some problems when I ran renovate manually the other day from this branch. it turned out to be an issue with the node-version. I'm pretty sure, this is unrelated to this branch, I just thought I mention it here:

locally I use [avn](https://github.com/wbyoung/avn) with its plugin [avn-nvm](https://github.com/wbyoung/avn-nvm) to handle different versions. So i started the renovate job using node `8.9.1`. However, the childprocesses for lockfile generation used the system default, which was `8.8.1`. This lead to an ChildProcessError during lockfile generation (also lead to an complete `npm install` in that childprocess instead of `package-lock-only`). Only after setting the system default to `8.9.1` everything worked as expected

I'm unsure
- if node `8.8.1` is broken
- if it is the different versions in processes (different context in some way)
- if I can grasp the problem at all 🤓

maybe you ran into issues like that before?

In real life, I guess this will not be an issue because the node version should be fixed on a server/cloud environment.

**Comment by rarkins** [comment]
@ctaepper I haven't seen that problem myself, although it's a little related to something I'm planning in #1420.

Previously:
- Users needed to have version x of `npm` and version y of `yarn` and you can guess it caused problems every now and then. It was worse before `yarn` reached `v1.0.0` and stabilised

Currently:
- `renovate` embeds `npm`, `yarn` and `pnpm` as `dependencies`, so we know which exact versions to run

Future:
- Stop including `npm`, `yarn` and `pnpm` as `dependencies`
- Instead of `child_process`, invoke `docker run` containers instead
- Publish an "official" versioned image like `@renovate/utils` that's used
- Allows us to control and call binaries that are non-npm installable, whether that's `git`, `composer`, `pip`, etc
- Still allow an option for people to fall back to globally installed binaries if they really want (e.g. if only using `npm`, skip the need for Docker)

When working more on this `lerna bootstrap` branch I realised that `lerna` is likely calling the *globally installed* `npm`/`yarn` on our local systems, and not the "bundled" ones. I don't think that's an unfair burden for now, but it's another example of why moving these all to a predictable Docker container will be good.

**Comment by ctaepper** [comment]
> renovate embeds npm, yarn and pnpm as dependencies, so we know which exact versions to run

hm, now that you mention that explicitly (I was kind of aware of that before), I remember a huge bug we ran into in the past:
- project was using node8 / npm5 
- we used some package to help executing npm scripts (can't remember exactly what it was)
- this package itsself had a dependency to npm, sadly some v2.x
- using npm5 to install the project, it flatted the npm 2.x dependecy into node_modules
- which resulted in any `scripts` in `package.json` which used npm to run other scripts (eg `"build": "npm run build:frontend && npm run build:backend"`), did not use the system- or project default, but the "old" version in the node_modules folder, because thats the way node_modules binaries are handled

this means, renovate causes the same potential problem. currently the dependency is fixed to 5.6.0. however, if my project would be relying on some older version, all run-scripts including `npm run ` will always use 5.6.0 and i can't do anything about it

**Comment by rarkins** [comment]
In theory it might be similar, but in practice I don't think it is. See, we only need to call `npm` or `lerna` if there is a `package-lock.json`, which implies `npm@5`. And npm have declared the lockfiles to be compatible in 5.x and also prefer you always using the latest anyway. If there was an `npm@6` with incompatible lock file then we'd need to add handling for that, but not now.

**Comment by rarkins** [comment]
@ctaepper I refactored this branch a bit to allow for non-workspaces `yarn` as well. Could you test this branch again to make sure it's still working for you, before I merge?

**Comment by ctaepper** [comment]
ran without errors!

**Pull request closed without merging**